solr/core/src/java/org/apache/solr/cloud/ZkController.java

```
@@ -2375,21 +2375,32 @@ public void rejoinShardLeaderElection(SolrParams params) {
-      if (roles == null) return;
-      @SuppressWarnings({"rawtypes"})
-      List nodeList = (List) roles.get("overseer");
-      if (nodeList == null) return;
-      if (nodeList.contains(getNodeName())) {
-        ZkNodeProps props = new ZkNodeProps(Overseer.QUEUE_OPERATION, CollectionParams.CollectionAction.ADDROLE.toString().toLowerCase(Locale.ROOT),
-            "node", getNodeName(),
-            "role", "overseer");
-        log.info("Going to add role {} ", props);
-        getOverseerCollectionQueue().offer(Utils.toJSON(props));
+      String s = System.getProperty("overseer.node");
+      Boolean isOverseer = s == null ? null : Boolean.parseBoolean(s);
+      byte[] data = null;
+      try {
+        data = zkClient.getData(ZkStateReader.ROLES, null, new Stat(), true);
+      } catch (NoNodeException e) {
+      }
+      @SuppressWarnings("unchecked")
+      Map<String, Object> roles = data == null ? Collections.emptyMap() : (Map<String, Object>) Utils.fromJSON(data);
+      List<String> overseerDesignates = (List<String>) roles.getOrDefault("overseer", Collections.emptyList());
+      CollectionParams.CollectionAction cmd = null;
+      if (isOverseer == null) {
+        if (overseerDesignates.contains(getNodeName())) cmd = CollectionParams.CollectionAction.ADDROLE;
+      } else {
+        if (isOverseer == Boolean.TRUE) {
+          cmd = CollectionParams.CollectionAction.ADDROLE;
+        } else if (isOverseer == Boolean.FALSE) {
+          if (overseerDesignates.contains(getNodeName())) cmd = CollectionParams.CollectionAction.REMOVEROLE;
+        }
       }
+      if (cmd == null) return;
+      ZkNodeProps props = new ZkNodeProps(Overseer.QUEUE_OPERATION, cmd.toLower(),
+          "node", getNodeName(),
+          "role", "overseer");
+      log.info("Going to add/remove role {} ", props);
+      getOverseerCollectionQueue().offer(Utils.toJSON(props));
     } catch (NoNodeException nne) {
       return;
```

if node is marked for overseer then for any exception should we just shutdown corecontainer?

---

yeah, that error will propagate up

---

looks like we are eating exception here?
```
} catch (NoNodeException nne) {
      return;
    } catch (Exception e) {
      log.warn("could not read the overseer designate ", e);
    }
```